src/cockpit/channel.py

```diff
@@ -433,31 +433,38 @@ class AsyncChannel(Channel):
     """
 
     # Receive-side flow control: intermix pings and data in the queue and reply
-    # to pings as we dequeue them.  This is a buffer: since we need to handle
-    # do_data() without blocking, we have no choice.
-    receive_queue = None
+    # to pings as we dequeue them.  EOF is None.  This is a buffer: since we
+    # need to handle do_data() without blocking, we have no choice.
+    receive_queue: 'asyncio.Queue[bytes | JsonObject | None]'
 
     # Send-side flow control
     write_waiter = None
 
-    async def run(self, options):
+    async def run(self, options: JsonObject) -> None:
         raise NotImplementedError
 
-    async def run_wrapper(self, options):
+    async def run_wrapper(self, options: JsonObject) -> None:
         try:
             await self.run(options)
             self.close()
         except ChannelError as exc:
             self.close(exc.attrs)
 
-    async def read(self):
+    async def read(self) -> 'bytes | None':
+        # Three possibilities for what we'll find:
+        #  - None (EOF) → return b''
+        #  - a ping → send a pong
+        #  - bytes → return it
         while True:
             item = await self.receive_queue.get()
-            if isinstance(item, bytes):
+            if item is None:
+                return None
+            if isinstance(item, Mapping):
```

Comment:
so this is a "map-ping", right?

:grin:

Reply:
Without look into the code, what does the `ping` Mapping look like?

Reply:
It just happens to be the only time we shove a dict into the queue.  Here's the other end:

```python
    def do_ping(self, message: JsonObject) -> None:
        self.receive_queue.put_nowait(message)
```

so if we find a dict in there, we can be pretty sure that it's a ping and we should pong it...